node/src/components/block_proposer.rs

```diff
@@ -402,32 +379,61 @@ where
                     .set_timeout(PRUNE_INTERVAL)
                     .event(|_| Event::BufferPrune);
             }
-            Event::Request(BlockProposerRequest::ListForInclusion {
-                current_instant,
-                past_blocks,
-                responder,
-            }) => {
-                return self.get_chainspec(effect_builder, current_instant, past_blocks, responder);
+            Event::Request(BlockProposerRequest::ListForInclusion(request)) => {
+                match request.last_finalized_block {
+                    Some(block_hash) if !self.state.finalized.contains_key(&block_hash) => {
+                        self.state
+                            .request_queue
+                            .entry(block_hash)
+                            .or_insert_with(Vec::new)
+                            .push(request);
+                    }
+                    _ => {
+                        return self.get_chainspec(effect_builder, request);
+                    }
+                }
             }
             Event::Buffer { hash, header } => self.add_deploy(Timestamp::now(), hash, *header),
-            Event::ProposedProtoBlock(block) => {
+            Event::FinalizedProtoBlock { block, parent } => {
                 let (hash, deploys, _) = block.destructure();
-                self.added_block(hash, deploys)
+                match parent {
+                    Some(parent_hash) if !self.state.finalized.contains_key(&parent_hash) => {
+                        self.state
+                            .finalization_queue
+                            .insert(parent_hash, (hash, deploys));
+                    }
+                    _ => {
```

Comment:
```suggestion
                    None | Some(_known_parent_hash) => {
```
(Also a matter of taste; but I'd prefer to be explicit.)